Jellyfin.Server.Implementations/Devices/DeviceManager.cs

```diff
@@ -0,0 +1,243 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Jellyfin.Data.Entities;
+using Jellyfin.Data.Entities.Security;
+using Jellyfin.Data.Enums;
+using Jellyfin.Data.Events;
+using Jellyfin.Data.Queries;
+using MediaBrowser.Controller.Devices;
+using MediaBrowser.Controller.Library;
+using MediaBrowser.Model.Devices;
+using MediaBrowser.Model.Querying;
+using MediaBrowser.Model.Session;
+using Microsoft.EntityFrameworkCore;
+
+namespace Jellyfin.Server.Implementations.Devices
+{
+    /// <summary>
+    /// Manages the creation, updating, and retrieval of devices.
+    /// </summary>
+    public class DeviceManager : IDeviceManager
+    {
+        private readonly JellyfinDbProvider _dbProvider;
+        private readonly IUserManager _userManager;
+        private readonly ConcurrentDictionary<string, ClientCapabilities> _capabilitiesMap = new ();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeviceManager"/> class.
+        /// </summary>
+        /// <param name="dbProvider">The database provider.</param>
+        /// <param name="userManager">The user manager.</param>
+        public DeviceManager(JellyfinDbProvider dbProvider, IUserManager userManager)
+        {
+            _dbProvider = dbProvider;
+            _userManager = userManager;
+        }
+
+        /// <inheritdoc />
+        public event EventHandler<GenericEventArgs<Tuple<string, DeviceOptions>>>? DeviceOptionsUpdated;
+
+        /// <inheritdoc />
+        public void SaveCapabilities(string deviceId, ClientCapabilities capabilities)
+        {
+            _capabilitiesMap[deviceId] = capabilities;
+        }
+
+        /// <inheritdoc />
+        public async Task UpdateDeviceOptions(string deviceId, string deviceName)
+        {
+            await using var dbContext = _dbProvider.CreateContext();
+            var deviceOptions = await dbContext.DeviceOptions.AsQueryable().FirstOrDefaultAsync(dev => dev.DeviceId == deviceId).ConfigureAwait(false);
+            if (deviceOptions == null)
+            {
+                deviceOptions = new DeviceOptions(deviceId);
+                dbContext.DeviceOptions.Add(deviceOptions);
```

Comment:
It's been a minute since I dabbled with EF, but can't we add the DeviceOptions regardless and only update the customname property? Ie. without checking it exists first

Reply:
(All of this is going off memory so it may be slightly wrong) We only want one device options to exist per device id. Since that id is not the primary key, doing so would always add a new entity and clash with the unique index on the device id field.

Reply:
Shouldn't it be the PK? We seem to be using it as such.

Reply:
Maybe? I made it this way to align with the other entities we have, but this can be changed (Maybe for 11 as well?)